src/main/scala/io/iohk/ethereum/blockchain/sync/regular/HeadersFetcher.scala

```diff
@@ -0,0 +1,81 @@
+package io.iohk.ethereum.blockchain.sync.regular
+import akka.actor.typed.{ActorRef, Behavior}
+import akka.actor.typed.scaladsl.{AbstractBehavior, ActorContext, Behaviors}
+import akka.actor.{ActorRef => ClassicActorRef}
+import io.iohk.ethereum.blockchain.sync.PeersClient.{BestPeer, Request}
+import io.iohk.ethereum.blockchain.sync.regular.BlockFetcher.FetchCommand
+import io.iohk.ethereum.blockchain.sync.regular.HeadersFetcher.HeadersFetcherCommand
+import io.iohk.ethereum.network.Peer
+import io.iohk.ethereum.network.p2p.Message
+import io.iohk.ethereum.network.p2p.messages.PV62.{BlockHeaders, GetBlockHeaders}
+import io.iohk.ethereum.utils.Config.SyncConfig
+import monix.eval.Task
+import monix.execution.Scheduler
+import org.slf4j.Logger
+
+import scala.util.{Failure, Success}
+
+class HeadersFetcher(
+    val peersClient: ClassicActorRef,
+    val syncConfig: SyncConfig,
+    val supervisor: ActorRef[FetchCommand],
+    context: ActorContext[HeadersFetcher.HeadersFetcherCommand]
+) extends AbstractBehavior[HeadersFetcher.HeadersFetcherCommand](context)
+    with FetchRequest[HeadersFetcherCommand] {
+
+  override val log: Logger = context.log
+  implicit val ec: Scheduler = Scheduler(context.executionContext)
+
+  import HeadersFetcher._
+
+  override def makeAdaptedMessage[T <: Message](peer: Peer, msg: T): HeadersFetcherCommand = AdaptedMessage(peer, msg)
+
+  override def onMessage(message: HeadersFetcherCommand): Behavior[HeadersFetcherCommand] =
+    message match {
+      case FetchHeaders(blockNumber: BigInt, amount: BigInt) =>
+        log.debug("Start fetching headers from block {}", blockNumber)
+        requestHeaders(blockNumber, amount)
+        Behaviors.same
+      case AdaptedMessage(_, BlockHeaders(headers)) =>
+        log.debug("Fetched {} headers starting from block {}", headers.size, headers.headOption.map(_.number))
+        supervisor ! BlockFetcher.ReceivedHeaders(headers)
+        Behaviors.same
+      case HeadersFetcher.RetryHeadersRequest =>
+        supervisor ! BlockFetcher.RetryHeadersRequest
+        Behaviors.same
+      case _ => Behaviors.unhandled
```

Comment:
The whole idea of typed actor is that you are only able to receive messages that extend `HeadersFetcherCommand`, so why do you feel the need to add `case _ => Behaviors.unhandled`?

Reply:
it doesn't compile without it

Reply:
Because of `case class AdaptedMessage[T <: Message]`, I see now